### ISSUE TITLE
expose_stan_functions fails on functions postfixed with `_lp`

### ISSUE THREAD

**Issue by bgoodri** [stan-dev/rstan#550]
#### Summary:

`expose_stan_functions` does not work with functions that modify `target`, which worked previously

#### Description:

If you try, there is a compiler error

#### Reproducible Steps:

`expose_stan_functions` this

```
functions {
  void foo_lp() {
    target += normal_lpdf(0.5 | 0, 1);
  }
}
```

#### Current Output:

```
In file included from file7d8f62dde969.cpp:1:
In file included from /usr/local/lib/R/site-library/rstan/include/exporter.h:1:
In file included from /usr/local/lib/R/site-library/Rcpp/include/RcppCommon.h:168:
In file included from /usr/local/lib/R/site-library/Rcpp/include/Rcpp/as.h:25:
/usr/local/lib/R/site-library/Rcpp/include/Rcpp/internal/Exporter.h:31:28: error: no matching constructor for initialization of 'stan::math::accumulator<double>'
                    Exporter( SEXP x ) : t(x){}
                                         ^ ~
/usr/local/lib/R/site-library/Rcpp/include/Rcpp/as.h:87:41: note: in instantiation of member function 'Rcpp::traits::Exporter<stan::math::accumulator<double> >::Exporter' requested here
            ::Rcpp::traits::Exporter<T> exporter(x);
                                        ^
/usr/local/lib/R/site-library/Rcpp/include/Rcpp/as.h:152:26: note: in instantiation of function template specialization 'Rcpp::internal::as<stan::math::accumulator<double> >' requested here
        return internal::as<T>(x, typename traits::r_type_traits<T>::r_category());
                         ^
/usr/local/lib/R/site-library/Rcpp/include/Rcpp/InputParameter.h:46:49: note: in instantiation of function template specialization 'Rcpp::as<stan::math::accumulator<double> >' requested here
        ReferenceInputParameter(SEXP x_) : obj( as<T>(x_) ){}
                                                ^
file7d8f62dde969.cpp:75:77: note: in instantiation of member function 'Rcpp::ReferenceInputParameter<stan::math::accumulator<double> >::ReferenceInputParameter' requested here
    Rcpp::traits::input_parameter< stan::math::accumulator<double>& >::type lp_accum__(lp_accum__SEXP);
                                                                            ^
/usr/local/lib/R/site-library/StanHeaders/include/stan/math/prim/mat/fun/accumulator.hpp:25:7: note: candidate constructor (the implicit copy constructor) not viable: cannot convert argument of incomplete type 'SEXP' (aka 'SEXPREC *') to 'const stan::math::accumulator<double>' for 1st argument
class accumulator {
      ^
/usr/local/lib/R/site-library/StanHeaders/include/stan/math/prim/mat/fun/accumulator.hpp:33:3: note: candidate constructor not viable: requires 0 arguments, but 1 was provided
  accumulator() : buf_() {}
  ^
```

and the relevant part of the generated C++ file is

```
void foo_lp(double& lp__, stan::math::accumulator<double>& lp_accum__, std::ostream* pstream__);
RcppExport SEXP sourceCpp_1_foo_lp(SEXP lp__SEXP, SEXP lp_accum__SEXP, SEXP pstream__SEXP) {
BEGIN_RCPP
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< double& >::type lp__(lp__SEXP);
    Rcpp::traits::input_parameter< stan::math::accumulator<double>& >::type lp_accum__(lp_accum__SEXP);
    Rcpp::traits::input_parameter< std::ostream* >::type pstream__(pstream__SEXP);
    foo_lp(lp__, lp_accum__, pstream__);
    return R_NilValue;
END_RCPP
}
```

#### Expected Output:

None

#### RStan Version:

develop (2.18) branch from GitHub

#### R Version:

3.5

#### Operating System:

Debian

**Comment by bgoodri** [comment]
@sakrejda I don't know how to fix this

**Comment by sakrejda** [comment]
Whoops, I'll take a look at it, we should have it increment a target variable, yeah?

**Comment by bgoodri** [comment]
The `target` variable should be passed in as an argument with a default of
zero

**Comment by sakrejda** [comment]
Ok, I'll do it Monday

**Comment by sakrejda** [comment]
Should the target be incremented or should it just add to the return value?

**Comment by bgoodri** [comment]
I think incremented because those functions are often void

**Comment by bob-carpenter** [comment]
Definitely incremented.  The _lp functions increment the target value passed by reference.

**Comment by sakrejda** [comment]
This needs two things: 

1) @syclik would it be ok to add a constructor to [accumulator.hpp](https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/accumulator.hpp) that created an accumulator by taking a pointer to a buffer (not ownership b/c R still deletes it at the end)?  We need this so that it can be a proxy for an R vector.

2) I need to add to [exporter.h](https://github.com/stan-dev/rstan/blob/develop/rstan/rstan/inst/include/exporter.h) in rstan so that it constructs the accumulator when it's required as an input parameter, probably adding

```

namespace Rcpp {
template <> stan::math::accumulator as(SEXP x) {
  // make the proxy accumulator with x's memory in its vector
};

...

 namespace traits {
 
   template <>
    struct input_parameter<stan::math::acumulator&> {
      typedef typename Rcpp::ConstReferenceInputParameter<stan::math::accumulator> type ;

    };

}
}
```

If Dan's on board with going this route I can make PR's, I thought I could do this without changes in math but we can't construct a new buffer and have it magically updated in R.

**Comment by syclik** [comment]
I'm not exactly sure how that would work with accumulator. (not saying it
can't, I think I don't have enough context) It looks RAII-ish and I don't
know how to make it both be RAII and non at the same time. Almost seems
like it'd be easier to call a different class? (It's code... I realize it
can be done.)

Do you have the flexibility to call something else where you're proposing
changes? And how do you guarantee there's enough memory allocated?

**Comment by sakrejda** [comment]
I agree that it's awkward.  I think I could derive from stan::math::accumulator in rstan and override everything we need. Now that I write this, I realize Rcpp is basically using the class as a tag so I'm not sure we even need to use the accumulator internally (although we still want something that acts like it). 

It's a little more code but I don't think efficiency is the same kind of issue there (no autodiff to worry about) so it would be straightforward code.

**Comment by bgoodri** [comment]
Why does the code need to pass a `stan::math::accumulator`? Why can't it just pass `double(0)` and increment that?

**Comment by sakrejda** [comment]
The signature R gets is determined by what Stan generates for standalone functions so the approaches I describe just deal with that.   A third option would be for stanc to generate a signature that takes a `double&` and plumb that through.   One thing I have not checked is how the accumulator is expected to behave: would a double work?

**Comment by bgoodri** [comment]
The generated code has stuff like

lp_accum__.add(normal_log(0.5,0,1));

so we probably have to figure out how to work with an accumulator.

**Comment by sakrejda** [comment]
That's what I guess, do you see any reason we can't derive from stan::math::accumulator and override the .add method?

**Comment by bgoodri** [comment]
Not that I know of. But can't we just put something into
expose_stan_functions_hacks() to get around it, like deleting the
accumulator and changing lp_accum__.add to lp__ += ?

**Comment by bgoodri** [comment]
Or, probably better, why doesn't the wrapper function construct a
stan::math::accumulator<double> internally and pass that to the function
inside the model's namespace?

**Comment by sakrejda** [comment]
I'm inclined to avoid hacks because there are plenty of clean ways to do this that will generalize to python.   It's just a question of which one to pick.  If you need this for a release now more hacks is fine but I'm not sure easier.   I could also expose the accumulator the way we did the rng so it could just be passed in, same kind a code would go into exporter.h

**Comment by bgoodri** [comment]
If the main concern is adding it to Python, then I think definitely the way
to go would be to have the wrapper functions construct the things that need
to be passed to the function inside the model's namespace but do not have a
native representation in the interface.

**Comment by bob-carpenter** [comment]
I think @bgoodri's proposal to instantiate a `stan::math::accumulator<double>` is a clean approach.

Another option would be to add `+=` operator implementations for accumulators on LHS and rewrite the code that uses accumulators to use `+=` instead of `add`.  That way, a `double&` instantiation would work.

A third option would be a custom implementation of the accumulator interface based just on a double reference.

But what @bgoodri is suggesting still seems easiest.

**Comment by sakrejda** [comment]
Ok, will do,  it's exactly what we did with the RNG and the R/Rcpp get_rng functions.

**Comment by sakrejda** [comment]
@bgoodri  I made a [branch](https://github.com/stan-dev/rstan/tree/bugfix/issue-550-postfix-lp) but I haven't gotten through re-installing StanHeaders/rstan so I'll check that tomorrow and fix whatever bugs there are.

**Comment by bgoodri** [comment]
What is the benefit to having the accumulator exposed to R? What can you do
with it besides pass it to a user-defined function ending in `_lp`?

**Comment by sakrejda** [comment]
I missed adding that part.  The only other thing you want to be able to do is check it's value.   It has to live in R memory somehow (as a ptr is easiest) s.t. its value can be accessed.

**Comment by sakrejda** [comment]
It's the same as RNG but we want to be able to check its value so we can see the result of the function.

**Comment by sakrejda** [comment]
With the current code on the branch I get:

```

> 
> library(rstan)
> check_accumulator(ACC)
[1] 0
> expose_stan_functions('model.stan')
> foo_lp(0, ACC, OUT)
> check_accumulator(ACC)
[1] -1.043939
> 
> acc = get_accumulator()
> check_accumulator(acc)
[1] 0
> foo_lp(0, acc, OUT)
> check_accumulator(acc)
[1] -1.043939
```

So... that seems good enough for a bugfix.  I'm realizing none of these have much roxygen doc.  I can add that this weekend.  They could probably do with explicit classes and print methods (like the print on an accumulator should just dump out the value)

**Comment by bgoodri** [comment]
Can't the function just return target() ?

**Comment by sakrejda** [comment]
I don't know what "target()" is so I probably missed something?

**Comment by sakrejda** [comment]
Part of the problem is that I don't use the _lp functions much so I don't know what you expect that syntax to do in rstan.

**Comment by bgoodri** [comment]
In the Stan function, you can do

real foo_lp(...) {
  target += something;
  return target();
}

and it yields the accumulated value.

**Comment by sakrejda** [comment]
I see, that's yet another signature,  it should work with the current exporter.h code, did you try one?

**Comment by sakrejda** [comment]
I'm on phone rn

**Comment by sakrejda** [comment]
Longer term we could change how standalone functions are generated in `stan-dev/stan` so that they always return target and don't create an accumulator at all.  We already do some special stuff for them.

**Comment by bgoodri** [comment]
I haven't tried your way yet. But what @bob-carpenter and I were saying is that the C++ wrapper could pass a `stan::math::accumulator` to the function in the model's namespace and then we don't have to expose anything on the interface side that is not usable in the interface. That is how I fixed it yesterday in
https://github.com/stan-dev/rstan/commit/dabdbfd9bbfe524ce7c0854aecd7c0efa53fc901

**Comment by sakrejda** [comment]
That's fine, but why not tell me yesterday you'd fixed it? Did I miss something?  I only fixed it because you asked and gsub-on-code is not an approach I want to maintain.  When rstan did more grep/gsub previously I kept having failures when I was trying to get my own models done!

**Comment by bgoodri** [comment]
Well, I think for 2.19 we should have the parser generate the code we want
so that we don't have to gsub (as much). But if everyone agrees that we
should construct the `stan::math::accumulator` on the inside, then it would
be a (minor) problem to introduce the accumulator in the interfaces in 2.18
and then take it away it 2.19.

**Comment by sakrejda** [comment]
Ok, that's fine with me.   Might want to check if the changes we're expecting to make in the parser are plausible.

**Comment by bgoodri** [comment]
Plausible @bob-carpenter ?

**Comment by bob-carpenter** [comment]
I'm not sure which of the topics you're asking about.  

Did you want some kind of change to the parser?

**Comment by bgoodri** [comment]
Having the parser generate

```
void
foo_lp(double& lp__, std::ostream* pstream__ = nullptr){
stan::math::accumulator<double> lp_accum__; // this line instead of it
being part of the signature
model_namespace::foo_lp<double, stan::math::accumulator<double> >(lp__,
lp_accum__, pstream__);
}
```
from
```
functions {
  void foo_lp() {
    target += normal_lpdf(0.5 | 0, 1);
  }
}
```

**Comment by sakrejda** [comment]
And also always adding `return target();`, right?

**Comment by bgoodri** [comment]
I think that should be up to the user.

**Comment by sakrejda** [comment]
That would mean you couldn't test functions that just rely on the accumulator from R whereas with my approach they both work.

**Comment by bgoodri** [comment]
If you want to test them or use the value from R, then you just return
target(); We do that a bunch in rstanarm.

**Comment by sakrejda** [comment]
Functions might return something else already and it's error-prone to edit a bunch of functions to test them so I don't like that answer.  This is your show so ¯\_(ツ)_/¯

**Comment by bgoodri** [comment]
This is breaking a bunch of tests. If you call multiple functions ending in `_lp` (or the same function more than once), you get the accumulated value of `target()` on the second call unless you call it with `lp_accum__ = get_accumulator()`.

**Comment by sakrejda** [comment]
Ok, I missed that in the test interface.   That's expected behavior since with an accumulator all that's guaranteed is the increment.   That's fixable,  and I'll take a look at setting default parameter values so this all disappears from the user's perspective.

**Comment by bgoodri** [comment]
Do you mean a default value for `get_accumulator` that the accumulator is set to?

**Comment by sakrejda** [comment]
Yes,  unfortunately it probably involves a PR to Rcpp since they currently hard code default value parsing.

**Comment by bgoodri** [comment]
Well, I am not sure how open they would be to that. But we have to gsub the
C++ code in the meantime.

I don't think the idea of an accumulator even makes much sense outside the
log_prob context, much less in the R or Python context.

**Comment by sakrejda** [comment]
The current pr works without any more gsub or default values,  you'd just have to pass all the requires args Stan requires

**Comment by bgoodri** [comment]
But it breaks tests in other R packages and changes the behavior of
expose_stan_functions() without being readily apparent to the person
calling the function in order to do something that is completely foreign in
the R / Python workflow.

**Comment by sakrejda** [comment]
Sure, if somebody is relying on undocumented text munging on c++ code sometimes their code will need updating. They should've used an explicit return to begin with.   As I said elsewhere I don't object to the code munging, I just don't want to take responsibility for maintaining the code munging.

**Comment by bob-carpenter** [comment]
RNGs maintain state, but I agree that it's unusual behavior for R and that we should maintain existing behavior.  

What is the existing behavior for _lp functions?  Is there a way to recover the target increment or is it just hidden?

**Comment by bgoodri** [comment]
The existing, and documented, behavior is that the user-defined function
ending in _lp is passed (by value) an initial value of lp__ (which defaults
to 0) that gets incremented from there. If users want to chain those
together, then they need to return target(); from the first function call
and pass the returned value to the lp__ argument of the next function call.

**Comment by bob-carpenter** [comment]
I found this doc, but couldn't follow it.  

> LP functions
If a user-defined Stan function ends in `_lp`, then it can modify the log-probability used by Stan to evaluate Metropolis proposals or as an objective function for optimization. When exposing such functions to R, a lp__ argument will be added to the formals. This `lp__` argument defaults to zero, but a double precision scalar may be passed to this argument when the function is called from R. Such a user-defined Stan function can terminate with return` target()`; or can execute `print(target())`; to verify that the calculation is correct.

I don't see from this how I'd call the exposed function, because I dont know what "formals" are.  I see there are "formal arguments" but I don't know what other kind there are (are the `...` not formal?).

Is `lp__` the name of an argument, as in I'd use `lp__ = x` to pass in a value?  

I take it the `target()` has to be inside the Stan program?  I didn't see doc for it as an RStan function.

And I take it there's no persistence of `target()` across multiple calls the way there is in a Stan program.

I take it the argument being passed is passed by value (I don't even know if there is pass-by-reference in R) so that it doesn't get modified the way the actual Stan function works?

**Comment by bgoodri** [comment]
The answer to all of those is "yes". And the formals of foo_lp are the
arguments to the function lp. There are no informals.
